dotCMS/src/main/java/com/dotcms/filters/interceptor/saml/SamlWebInterceptor.java

```diff
@@ -0,0 +1,551 @@
+package com.dotcms.filters.interceptor.saml;
+
+import com.dotcms.cms.login.LoginServiceAPI;
+import com.dotcms.filters.interceptor.Result;
+import com.dotcms.filters.interceptor.WebInterceptor;
+import com.dotcms.saml.DotSamlProxyFactory;
+import com.dotcms.saml.IdentityProviderConfiguration;
+import com.dotcms.saml.IdentityProviderConfigurationFactory;
+import com.dotcms.saml.SamlAuthenticationService;
+import com.dotcms.saml.SamlConfigurationService;
+import com.dotcms.saml.SamlName;
+import com.dotcms.security.apps.AppsAPI;
+import com.dotcms.util.security.Encryptor;
+import com.dotcms.util.security.EncryptorFactory;
+import com.dotmarketing.beans.Host;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.NoSuchUserException;
+import com.dotmarketing.business.UserAPI;
+import com.dotmarketing.business.web.HostWebAPI;
+import com.dotmarketing.business.web.WebAPILocator;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.RegEX;
+import com.dotmarketing.util.SecurityLogger;
+import com.dotmarketing.util.UtilMethods;
+import com.dotmarketing.util.WebKeys;
+import com.liferay.portal.auth.PrincipalThreadLocal;
+import com.liferay.portal.model.User;
+import com.liferay.portal.servlet.PortletSessionPool;
+import io.vavr.control.Try;
+import org.apache.commons.lang.StringUtils;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.dotcms.saml.DotSamlConstants.SAML_USER_ID;
+
+/**
+ * This interceptor encapsulates the logic for Saml
+ * Basically if there is any configuration set on dot apps portlet for "app-saml-config" and there is a configuration for the current host,
+ * the interceptor will try to autologin the user (if there is any saml user id on the session) otherwise will redirect to the IDP login.
+ * In addition if there is any saml host configuration for logout and the interceptor realized it is a logout request, will do the logout on dotCMS
+ * in addition to the logout on the idp if the configuration requires.
+ * @author jsanca
+ */
+public class SamlWebInterceptor implements WebInterceptor {
+
+    public static final String REFERRER_PARAMETER_KEY = "referrer";
+    public static final String ORIGINAL_REQUEST       = "original_request";
+
+    protected final Encryptor       encryptor;
+    protected final LoginServiceAPI loginService;
+    protected final UserAPI         userAPI;
+    protected final HostWebAPI      hostWebAPI;
+    protected final AppsAPI         appsAPI;
+    protected final SamlWebUtils    samlWebUtils;
+    protected final IdentityProviderConfigurationFactory identityProviderConfigurationFactory;
+    protected volatile SamlConfigurationService samlConfigurationService = null;
```

Comment:
Codacy found an issue: [Avoid using redundant field initializer for 'samlConfigurationService'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)